### PR TITLE
Replace - with _ for creating hive tables

### PR DESCRIPTION
**Pull request by sreevatsanraman**
* Hive-0.12 supports only _ other than alphanumeric - replacing '-' with '_'
* will document this in another PR

### CODE REVIEW

**Comment by sreevatsanraman** [comment]
Build: http://builds.cask.co/browse/CDAP-DUT849-1

File: cdap-explore/src/main/java/co/cask/cdap/explore/executor/ExploreExecutorHttpHandler.java [hunk 1]
```diff
@@ -434,7 +434,7 @@ private static String getStreamTableName(Id.Stream streamId) {
   private static String getHiveTableName(String name) {
     // Instance name is like cdap.user.my_table.
     // For now replace . with _ since Hive tables cannot have . in them.
-    return name.replaceAll("\\.", "_").toLowerCase();
+    return name.replaceAll("\\.","-").replaceAll("-", "_").toLowerCase();
```
Comment (poornachandra) [inline cdap-explore/src/main/java/co/cask/cdap/explore/executor/ExploreExecutorHttpHandler.java]: It should be `name.replaceAll("\\.","_")`, right?

**Comment by anwar6953** [comment]
This isn't a good idea?
If you're telling users that stream name can have both '-' and '_' in the stream name, then these two streams should be able to coexist and fully functional:
`sree-stream` and `sree_stream`.

Two alternatives:
-encode '-' with some character that works for hive, but not for stream names
-completely disallow '-' from stream names

File: cdap-explore/src/test/java/co/cask/cdap/explore/service/HiveExploreServiceStreamTest.java [hunk 2]
```diff
@@ -155,6 +154,20 @@ public void testSelectAndFilterQueryOnStream() throws Exception {
   }
 
   @Test
+  public void testStreamNameWithHyphen() throws Exception {
```
Comment (poornachandra) [inline cdap-explore/src/test/java/co/cask/cdap/explore/service/HiveExploreServiceStreamTest.java]: Can you add one for `.` too?

**Comment by jwang47** [comment]
I agree with @anwar6953. Also, don't we already have the issue of multiple streams coexisting? For example, I believe Hive table names are case-insensitive and I know that if we create a stream like "logEventStream", then the Hive table for that stream will be "logeventstream".

File: cdap-explore/src/test/java/co/cask/cdap/explore/service/HiveExploreServiceStreamTest.java (hunk 2 above)
Comment (sreevatsanraman) [inline cdap-explore/src/test/java/co/cask/cdap/explore/service/HiveExploreServiceStreamTest.java]: Streams doesn't allow '.' 

Added tests for '.' and '_' in dataset code.

**Comment by sreevatsanraman** [comment]
@anwar6953 @jwang47 I agree with you guys. But this is probably the best given the constraints. 
Hive does support only alphanumeric characters and underscore. 

We are going to document this so that users can choose names appropriately.

**Comment by anwar6953** [comment]
How about simply disallowing '-' from stream names?
And documenting that?

**Comment by sreevatsanraman** [comment]
We considered that possibility and discussed with @ghelmling  and @poornachandra . But we did not to restrict ourselves because of limitations in Hive. The various system we integrate might have different constraints, we might not want to surface all these limitations to higher level abstractions we have.

**Comment by anwar6953** [comment]
Well, this way, we can simply say we are not restricting the user, but CDAP is still functionally restricting the user (because CDAP offers hive querying of streams).
(The users aren't restricted according to the docs, but they will find out the hard way that certain features are broken, which is worse).

**Comment by sreevatsanraman** [comment]
Hive exploration is just a part of CDAP and we might not have this restriction in future version of hive. I don't think it is that critical.

**Comment by anwar6953** [comment]
Alright, lets keep this open then:
https://issues.cask.co/browse/CDAP-1296

**Comment by sreevatsanraman** [comment]
@anwar6953 - I will open a new JIRA.

**Comment by anwar6953** [comment]
ok

File: cdap-explore/src/main/java/co/cask/cdap/explore/executor/ExploreExecutorHttpHandler.java [hunk 3]
```diff
@@ -434,7 +434,7 @@ private static String getStreamTableName(Id.Stream streamId) {
   private static String getHiveTableName(String name) {
     // Instance name is like cdap.user.my_table.
     // For now replace . with _ since Hive tables cannot have . in them.
```
Comment (poornachandra) [inline cdap-explore/src/main/java/co/cask/cdap/explore/executor/ExploreExecutorHttpHandler.java]: Update comment

**Comment by poornachandra** [comment]
LGTM

**Pull request merged**